services/agent/src/simcore_service_agent/modules/task_monitor.py

```diff
@@ -0,0 +1,140 @@
+import asyncio
+import logging
+import traceback
+from collections import deque
+from contextlib import suppress
+from dataclasses import dataclass
+from typing import Any, Awaitable, Callable, Coroutine, Final, Optional
+
+from fastapi import FastAPI
+from pydantic import PositiveFloat, PositiveInt
+
+from ..core.settings import ApplicationSettings
+from .volumes_cleanup import backup_and_remove_volumes
+from servicelib.logging_utils import log_context
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_TASK_WAIT_ON_ERROR: Final[PositiveInt] = 10
+
+
+@dataclass
+class _TaskData:
+    callable: Callable
+    args: Any
+    repeat_interval_s: Optional[PositiveFloat]
+
+
+async def _task_runner(task_data: _TaskData) -> None:
+    with log_context(logger, logging.INFO, msg=f"'{task_data.callable.__name__}'"):
+        while True:
+            coroutine: Coroutine = task_data.callable(*task_data.args)
+            try:
+                await coroutine
+            except Exception as e:  # pylint: disable=broad-except
+                logger.error(
+                    "Had an error while running '%s':\n%s",
+                    coroutine.__name__,
+                    "\n".join(traceback.format_tb(e.__traceback__)),
+                )
+
+            if task_data.repeat_interval_s is None:
+                logger.warning(
+                    "Unexpected termination of '%s'; it will be restarted",
+                    coroutine.__name__,
+                )
+
+            logger.info(
+                "Will run '%s' again in %s seconds",
+                coroutine.__name__,
+                task_data.repeat_interval_s,
+            )
+            await asyncio.sleep(
+                DEFAULT_TASK_WAIT_ON_ERROR
+                if task_data.repeat_interval_s is None
+                else task_data.repeat_interval_s
+            )
+
+
+class TaskMonitor:
+    def __init__(self):
+        self._running: bool = False
+        self._tasks: set[asyncio.Task] = set()
+
+        self._to_start: dict[str, _TaskData] = {}
+
+    @property
+    def is_running(self) -> bool:
+        return self._running
+
+    def register_job(
+        self,
+        target: Callable,
+        *args: Any,
+        repeat_interval_s: Optional[PositiveFloat] = None,
+    ) -> None:
+        if self._running:
+            raise RecursionError(
+                "Cannot add more tasks, monitor already running with: "
+                f"{[x.get_name() for x in self._tasks]}"
+            )
+
+        if target.__name__ in self._to_start:
+            raise RuntimeError(f"{target.__name__} is already registered")
+
+        self._to_start[target.__name__] = _TaskData(target, args, repeat_interval_s)
+
+    async def start(self) -> None:
+        self._running = True
```

Comment:
ok, fundamentally this does not change much isn't it? it is going to be true whatever happens in the middle of the background task.

Reply:
I think the name created some confusion. I've renamed it to `_was_started`. It is only relative to the `TaskManager` and not to the `jobs` it's suppose to keep track of.

To answer your question about health checks. I will be adding a new PR, where metrics for running tasks will be published by this service. The healthck will be derived from those. I'd like to create a metric keeping track of errors. For example: if more than X errors occur in the past Y runs, mark services as unhealthy.
Will discuss this further with @mrnicegyu11 

The current not so useful health check will be replaced.

Reply:
I still do not get the point. the healthcheck is there to allow docker to restart the service if it is malfunctioning. in your case, that would be if the background task is hanging somewhere or killed by an exception. Right now the only way to solve it is by manually restarting it (provided we detect it which is even another issue). Therefore the healthcheck is useless and is just generating traffic.